tests/phpunit/CRM/MembershipExtras/Hook/PostProcess/UpdateSubscriptionTest.php

```diff
@@ -0,0 +1,476 @@
+
+  /**
+   * Parameters for line items.
+   *
+   * @var array
+   */
+  private $lineItemsParams = [];
+
+  /**
+   * Parameters for first installment.
+   *
+   * @var array
+   */
+  private $contributionParams = [];
+
+  /**
+   * Data for the "Member Dues" financial type.
+   *
+   * @var array
+   */
+  private $memberDuesFinancialType = [];
+
+  /**
+   * Payment instrument ID for EFT.
+   *
+   * @var int
+   */
+  private $eftPaymentInstrumentID = 0;
+
+  /**
+   * ID of 'Pending' Contribution Status.
+   *
+   * @var int
+   */
+  private $contributionPendingStatusValue = 0;
+
+  /**
+   * Data for the default membership price set.
+   *
+   * @var array
+   */
+  private $defaultMembershipsPriceSet = [];
+
+  /**
+   * The form used to update recurring contributions.
+   *
+   * @var CRM_Contribute_Form_UpdateSubscription
+   */
+  private $updateSubscriptionForm;
+
+  /**
+   * Sets up common data required to run tests.
+   *
+   * @throws \CiviCRM_API3_Exception
+   */
+  public function setUp() {
+    $this->loadParameters();
+    $this->fabricateRequiredEntities();
+    $this->setUpDefaultPaymentPlanParameters();
+    $this->setUpUpdateSubscriptionForm();
+  }
+
+  /**
+   * Loads parameters required for the tests.
+   *
+   * @throws \CiviCRM_API3_Exception
+   */
+  private function loadParameters() {
+    $this->contributionPendingStatusValue = $this->getPendingContributionStatusValue();
+    $this->memberDuesFinancialType = $this->getMembershipDuesFinancialType();
+    $this->eftPaymentInstrumentID = $this->getEFTPaymentInstrumentID();
+    $this->defaultMembershipsPriceSet = $this->getDefaultPriceSet();
+  }
+
+  /**
+   * Obtains value for the 'Pending' contribution status option value.
+   *
+   * @return array
+   * @throws \CiviCRM_API3_Exception
+   */
+  private function getPendingContributionStatusValue() {
+    return civicrm_api3('OptionValue', 'getvalue', [
+      'return' => 'value',
+      'option_group_id' => 'contribution_status',
+      'name' => 'Pending',
+    ]);
+  }
+
+  /**
+   * Obtains default price set for memberships.
+   *
+   * @return array
+   * @throws \CiviCRM_API3_Exception
+   */
+  private function getDefaultPriceSet() {
+    $result = civicrm_api3(
+      'PriceSet', 'get', [
+      'sequential' => 1,
+      'name' => 'default_membership_type_amount',
+      'options' => ['limit' => 1],
+    ]);
+
+    if ($result['count'] > 0) {
+      return array_shift($result['values']);
+    }
+
+    return [];
+  }
+
+  /**
+   * Fabricates entities required for tests.
+   *
+   * @throws \CiviCRM_API3_Exception
+   */
+  private function fabricateRequiredEntities() {
+    $this->contact = ContactFabricator::fabricate();
+
+    $this->membershipType = MembershipTypeFabricator::fabricate(
+      [
+        'name' => 'Test Rolling Membership',
+        'period_type' => 'rolling',
+        'minimum_fee' => 120,
+      ],
+      TRUE
+    );
+    $priceField = PriceFieldFabricator::fabricate([
```

Comment:
Yeah, the MembershipType fabricator was built by Tunbola for other tests. Changing it would require updating all those test classes (MembershipTypeAmountProraterTest, MembershipTypeDurationCalculatorTest and MembershipTypeInstalmentAmountCalculatorTest), with a total of around 12 individual tests distributed among them. So it could take a while. Do you think we could update this after we merge in this PR?

@omarabuhussein , @erawat

Reply:
I am fine if that is done in separate PR, but we need to do it soon, like right after merging this PR.  don't want it to get overused in our tests or it will be a nightmare for us to change later.